docs/standard/io/pipelines.md

```diff
@@ -0,0 +1,342 @@
+  - "Pipelines"
+  - "Pipelines I/O"
+  - "I/O [.NET], Pipelines"
+author: rick-anderson
+ms.author: riande
+---
+# System.IO.Pipelines in .NET
+
+<xref:System.IO.Pipelines> is a new library that is designed to make it easier to do high-performance I/O in .NET. It’s a library targeting .NET Standard that works on all .NET implementations.
+
+<a name="solve"></a>
+
+## What problem does System.IO.Pipelines solve
+<!-- corner case doesn't MT (machine translate)   -->
+Apps that parse streaming data are composed of boilerplate code having many specialized and unusual code flows. The boilerplate and special case code is complex and difficult to maintain.
+
+`System.IO.Pipelines` was architected to:
+
+* Have high performance parsing streaming data.
+* Reduce code complexity.
+
+The following code is typical for a TCP server that receives line-delimited messages (delimited by `'\n'`) from a client:
+
+```csharp
+async Task ProcessLinesAsync(NetworkStream stream)
+{
+    var buffer = new byte[1024];
+    await stream.ReadAsync(buffer, 0, buffer.Length);
+    
+    // Process a single line from the buffer
+    ProcessLine(buffer);
+}
+```
+
+The preceding code has several problems:
+
+* The entire message (end of line) might not be received in a single call to `ReadAsync`.
+* It's ignoring the result of `stream.ReadAsync`. `stream.ReadAsync` returns how much data was read.
+* It doesn't handle the case where multiple lines are read in a single `ReadAsync` call.
+
+To fix the preceding problems, the following changes are required:
+
+* Buffer the incoming data until a new line is found.
+* Parse all the lines returned in the buffer.
+* It's possible that the line is bigger than 1 KB (1024 bytes). The code needs to resize the input buffer a complete line is found.
+
+  * If the buffer is resized, more buffer copies are made as longer lines appear in the input.
+  * To reduce wasted space, compact the buffer used for reading lines.
+
+* Consider using buffer pooling to avoid allocating memory repeatedly.
+* The following code address some of these problems:
+
+[!code-csharp[](~/samples/snippets/csharp/pipelines/ProcessLinesAsync.cs?name=snippet)]
+
+The previous code is complex and doesn't address all the problems identified. High-performance networking usually means writing very complex code to maximize performance. `System.IO.Pipelines` was designed to make writing this type of code easier.
+
+## Pipe
+
+The <xref:System.IO.Pipelines.Pipe> class can be used to create a `PipeWriter/PipeReader` pair. All data written into the `PipeWriter` is available in the `PipeReader`:
+
+[!code-csharp[](~/samples/snippets/csharp/pipelines/Pipe.cs?name=snippet2)]
+
+<a name="pbu"></a>
+
+### Pipe basic usage
+
+[!code-csharp[](~/samples/snippets/csharp/pipelines/Pipe.cs?name=snippet)]
+
+There are two loops:
+
+* `FillPipeAsync` reads from the `Socket` and writes to the `PipeWriter`.
+* `ReadPipeAsync` reads from the `PipeReader` and parses incoming lines.
```

Comment:
I noticed mixed usage of `Complete` and `CompleteAsync` in this doc. These two methods call `Complete`, but `CompleteAsync` is also mentioned in a few places.

Shouldn't we use `CompleteAsync` everywhere to encourage people to use it and avoid possible sync-over-async bugs?

Reply:
Agreed. We should deprecate Complete IMO.

Reply:
```
async Task FillPipeAsync(Socket socket, PipeWriter writer)
{
// code omitted
// Tell the PipeReader that there's no more data coming.
    writer.Complete();   // <--- deprecate Complete IMO.
}

async Task ReadPipeAsync(PipeReader reader)
{
// code omitted

    // Mark the PipeReader as complete.
    reader.Complete();                              // <-   deprecate Complete IMO.
}
```

I'm not allowed to modify code. If one of your are brave, can you update that at https://github.com/dotnet/samples/tree/master/snippets/csharp/pipelines ?

Reply:
Done. https://github.com/dotnet/samples/pull/1624

Reply:
Done. https://github.com/dotnet/samples/pull/1624